### PR TITLE
cache

### PR DESCRIPTION
**Pull request by gutorov**

### CODE REVIEW

**Review by Ikhsanov-Nail-95 — APPROVED** [review]
Все супер!

File: src/main/java/faang/school/achievement/cache/AchievementCache.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            achievementsByName.put(achievement.getTitle(), achievement);
+        });
+    }
+
+    public Optional<Achievement> getAchievement(String title) {
+        return Optional.ofNullable(achievementsByName.get(title));
+    }
+
+    public HashMap<String, Achievement> getAchievementsByName() {
+        return (HashMap<String, Achievement>) achievementsByName;
```
Comment (adilbeknurmakov) [inline src/main/java/faang/school/achievement/cache/AchievementCache.java:33]: I would return copy of hashmap then original. Reason -  Encapsulation / Immutability: return new HashMap<>(achievementsByName); Your hashmap here is like a cache it is bad practice if your cache it not immutable or threadsafe.
Or another may you could use concurrent hashmap

**Review by InnMish — APPROVED** [review]

File: src/main/java/faang/school/achievement/cache/AchievementCache.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        achievements.forEach(achievement -> {
+            achievementsByName.put(achievement.getTitle(), achievement);
+        });
+    }
+
+    public Optional<Achievement> getAchievement(String title) {
+        return Optional.ofNullable(achievementsByName.get(title));
+    }
+
+    public HashMap<String, Achievement> getAchievementsByName() {
```
Comment (adilbeknurmakov) [inline src/main/java/faang/school/achievement/cache/AchievementCache.java:32]: Or you have 2-nd option avoid using manual map - it is the best option. Use Spring @Cachebale annotation-Spring make caching for you in repository. To use it add @EnableCaching spring boot main class. This way with this annotation you can directly use achivementRepository. like this:  @Cacheable(value = "achievements", key = "#title") for method getAchievement. For method getAllAchievment you can do the same way

Comment (gutorov) [reply src/main/java/faang/school/achievement/cache/AchievementCache.java:32]: Sounds good, but I was asked to do with HashpMap . But I would rather try it, as you recommended

File: src/main/java/faang/school/achievement/cache/AchievementCache.java (hunk 1 above)
Comment (gutorov) [reply src/main/java/faang/school/achievement/cache/AchievementCache.java:33]: Sounds reasonable

**Pull request merged**